README.rst
=================================
Async Lutron Integration Protocol
=================================


.. image:: https://img.shields.io/pypi/v/aiolip.svg
        :target: https://pypi.python.org/pypi/aiolip

.. image:: https://img.shields.io/travis/bdraco/aiolip.svg
        :target: https://travis-ci.com/bdraco/aiolip

.. image:: https://readthedocs.org/projects/aiolip/badge/?version=latest
        :target: https://aiolip.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Async Lutron Integration Protocol


* Free software: Apache Software License 2.0
* Documentation: https://aiolip.readthedocs.io.


Example Usage
-------------

.. code-block:: python

        import asyncio
        import logging

        from aiolip import LIP
        from aiolip.data import LIPMode

        _LOGGER = logging.getLogger(__name__)


        async def main():
        lip = LIP()

        logging.basicConfig(level=logging.DEBUG)

        await lip.async_connect("192.168.209.70")

        def message(msg):
                _LOGGER.warning(msg)

        lip.subscribe(message)
        run_task = asyncio.create_task(lip.async_run())
        await run_task
        await lip.async_stop()

        if __name__ == "__main__":
        loop = asyncio.get_event_loop()
        loop.run_until_complete(main())

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

1.0.0 (2021-01-18)
------------------

* First release on PyPI.


test.py
import asyncio
import logging

from aiolip import LIP
from aiolip.data import LIPMode

_LOGGER = logging.getLogger(__name__)


async def main():
    lip = LIP()

    logging.basicConfig(level=logging.DEBUG)

    await lip.async_connect("192.168.209.70")

    def message(msg):
        _LOGGER.warning(msg)

    lip.subscribe(message)
    task = asyncio.create_task(lip.async_run())
    await lip.query(LIPMode.OUTPUT, 31, 1)
    #await lip.async_stop()
    await task

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())


tests/__init__.py
"""Unit test package for aiolip."""


tests/test_aiolip.py
#!/usr/bin/env python

"""Tests for `aiolip` package."""

import pytest

from aiolip import LIP


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_create(response):
    """Test to create the object."""
    LIP()


aiolip/__init__.py
import asyncio
import logging
import socket
import time

from .data import LIP_PROTOCOL_MODE_TO_LIPMODE, LIPMessage, LIPMode
from .exceptions import LIPConnectionStateError, LIPProtocolError
from .protocol import (
    CONNECT_TIMEOUT,
    LIP_ACTION_CHAR,
    LIP_EMPTY_RE,
    LIP_ERROR_RE,
    LIP_KEEP_ALIVE,
    LIP_KEEP_ALIVE_INTERVAL,
    LIP_KEEPALIVE_RE,
    LIP_PASSWORD,
    LIP_PORT,
    LIP_PROTOCOL_GNET,
    LIP_PROTOCOL_LOGIN,
    LIP_PROTOCOL_PASSWORD,
    LIP_QUERY_CHAR,
    LIP_READ_TIMEOUT,
    LIP_RESPONSE_RE,
    LIP_USERNAME,
    SOCKET_TIMEOUT,
    LIPConenctionState,
    LIPSocket,
)

_LOGGER = logging.getLogger(__name__)


class LIP:
    """Async class to speak LIP."""

    def __init__(self):
        """Create the LIP class."""
        self.connection_state = LIPConenctionState.NOT_CONNECTED
        self._lip = None
        self._host = None
        self._read_connect_lock = asyncio.Lock()
        self._disconnect_event = asyncio.Event()
        self._reconnecting_event = asyncio.Event()
        self._subscriptions = []
        self._last_keep_alive_response = None
        self._keep_alive_task = None
        self.loop = None

    async def async_connect(self, server_addr):
        """Connect to the bridge via LIP."""
        self.loop = asyncio.get_event_loop()

        if self.connection_state != LIPConenctionState.NOT_CONNECTED:
            raise LIPConnectionStateError

        self._disconnect_event.clear()

        try:
            await self._async_connect(server_addr)
        except asyncio.TimeoutError:
            _LOGGER.debug("Timed out while trying to connect to %s", server_addr)
            self.connection_state = LIPConenctionState.NOT_CONNECTED
            raise

    async def _async_connect(self, server_addr):
        """Make the connection."""
        _LOGGER.debug("Connecting to %s", server_addr)
        self.connection_state = LIPConenctionState.CONNECTING
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(
                server_addr,
                LIP_PORT,
                family=socket.AF_INET,
            ),
            timeout=CONNECT_TIMEOUT,
        )
        self._lip = LIPSocket(reader, writer)
        _verify_expected_response(
            await self._lip.async_readuntil(" "), LIP_PROTOCOL_LOGIN
        )
        await self._lip.async_write_command(LIP_USERNAME)
        _verify_expected_response(
            await self._lip.async_readuntil(" "), LIP_PROTOCOL_PASSWORD
        )
        await self._lip.async_write_command(LIP_PASSWORD)
        _verify_expected_response(
            await self._lip.async_readuntil(" "), LIP_PROTOCOL_GNET
        )
        self.connection_state = LIPConenctionState.CONNECTED
        self._host = server_addr
        self._reconnecting_event.clear()
        _LOGGER.debug("Connected to %s", server_addr)

    async def _async_disconnected(self):
        """Reconnect after disconnected."""
        if self._reconnecting_event.is_set():
            return

        if self._lip:
            self._lip.close()
            self._lip = None

        self._reconnecting_event.set()
        async with self._read_connect_lock:
            self.connection_state = LIPConenctionState.NOT_CONNECTED
            while not self._disconnect_event.is_set():
                try:
                    await self._async_connect(self._host)
                    return
                except asyncio.TimeoutError:
                    _LOGGER.debug(
                        "Timed out while trying to reconnect to %s", self._host
                    )
                    pass

    async def async_stop(self):
        """Disconnect from the bridge."""
        self._disconnect_event.set()
        if self._keep_alive_task:
            self._keep_alive_task.cancel()
            self._keep_alive_task = None
        self._lip.close()

    async def _async_keep_alive_or_reconnect(self):
        """Keep alive or reconnect."""
        connection_error = False
        try:
            await self._lip.async_write_command(LIP_KEEP_ALIVE)
        except (asyncio.TimeoutError, ConnectionResetError) as ex:
            _LOGGER.debug("Lutron bridge disconnected: %s", ex)
            connection_error = True

        if connection_error or self._last_keep_alive_response < time.time() - (
            LIP_KEEP_ALIVE_INTERVAL + SOCKET_TIMEOUT
        ):
            _LOGGER.debug("Lutron bridge keep alive timeout, reconnecting.")
            await self._async_disconnected()
            self._last_keep_alive_response = time.time()

    def _keepalive_watchdog(self):
        """Send keep alives."""
        if (
            self._disconnect_event.is_set()
            or self.connection_state != LIPConenctionState.CONNECTED
        ):
            return

        asyncio.create_task(self._async_keep_alive_or_reconnect())

        self._keep_alive_task = self.loop.call_later(
            LIP_KEEP_ALIVE_INTERVAL, self._keepalive_watchdog
        )

    async def async_run(self):
        """Start interacting with the bridge."""
        if self.connection_state != LIPConenctionState.CONNECTED:
            raise LIPConnectionStateError

        self._last_keep_alive_response = time.time()
        self._keepalive_watchdog()

        while not self._disconnect_event.is_set():
            await self._async_run_once()

    async def _async_run_once(self):
        """Process one message or event."""
        async with self._read_connect_lock:
            read_task = asyncio.create_task(self._lip.async_readline())

            _, pending = await asyncio.wait(
                (
                    read_task,
                    self._disconnect_event.wait(),
                    self._reconnecting_event.wait(),
                ),
                timeout=LIP_READ_TIMEOUT,
                return_when=asyncio.FIRST_COMPLETED,
            )
            for coro in pending:
                coro.cancel()

        if self._disconnect_event.is_set():
            _LOGGER.debug("Stopping run because of disconnect_event")
            return

        if isinstance(read_task.exception(), asyncio.TimeoutError) or isinstance(
            read_task.exception(), BrokenPipeError
        ):
            return

        self._process_message(read_task.result())

    def _process_message(self, response):
        """Process a lip message."""
        if response is not None and response.startswith(LIP_PROTOCOL_GNET):
            response = response[len(LIP_PROTOCOL_GNET) :]

        if not response or LIP_EMPTY_RE.match(response):
            return

        match = LIP_RESPONSE_RE.match(response)
        if match:
            try:
                message = LIPMessage(
                    LIP_PROTOCOL_MODE_TO_LIPMODE.get(match.group(1), LIPMode.UNKNOWN),
                    int(match.group(2)),
                    int(match.group(3)),
                    float(match.group(4)),
                )
                for callback in self._subscriptions:
                    callback(message)
            except ValueError:
                pass
        elif LIP_KEEPALIVE_RE.match(response):
            self._last_keep_alive_response = time.time()
        elif LIP_ERROR_RE.match(response):
            _LOGGER.error("Protocol Error: %s", response)
        else:
            _LOGGER.debug("Unknown lutron message: %s", response)

    async def query(self, mode, integration_id, action, *args):
        """Query the bridge."""
        await self._async_send_command(
            LIP_QUERY_CHAR, mode, integration_id, action, *args
        )

    async def action(self, mode, integration_id, action, *args):
        """Do an action on the bridge."""
        await self._async_send_command(
            LIP_ACTION_CHAR, mode, integration_id, action, *args
        )

    def subscribe(self, callback):
        """Subscribe to lip events."""

        def _unsub_callback():
            self._subscriptions.remove(callback)

        self._subscriptions.append(callback)
        return _unsub_callback

    async def _async_send_command(self, protocol_header, mode, *cmd):
        """Send a command."""
        if self.connection_state != LIPConenctionState.CONNECTED:
            raise LIPConnectionStateError

        assert isinstance(mode, LIPMode)

        request = ",".join([mode.name, *[str(key) for key in cmd]])
        await self._lip.async_write_command(f"{protocol_header}{request}")


def _verify_expected_response(received, expected):
    if not received.startswith(expected):
        raise LIPProtocolError(received, expected)


aiolip/protocol.py
import asyncio
from enum import Enum
import re

LIP_RESPONSE_RE = re.compile("~([A-Z]+),([0-9.]+),([0-9.]+),([0-9.]+)")
LIP_KEEPALIVE_RE = re.compile("~SYSTEM,")
LIP_EMPTY_RE = re.compile("^[\r\n]+$")
LIP_ERROR_RE = re.compile("~ERROR,")

LIP_PROTOCOL_LOGIN = "login: "
LIP_PROTOCOL_PASSWORD = "password: "
LIP_PROTOCOL_GNET = "GNET> "

LIP_USERNAME = "lutron"
LIP_PASSWORD = "integration"
LIP_PORT = 23

LIP_QUERY_CHAR = "?"
LIP_ACTION_CHAR = "#"

LIP_BUTTON_PRESS = 3
LIP_BUTTON_RELEASE = 4

LIP_KEEP_ALIVE = "?SYSTEM,10"

CONNECT_TIMEOUT = 10
SOCKET_TIMEOUT = 10

LIP_READ_TIMEOUT = 60
LIP_KEEP_ALIVE_INTERVAL = 60


class LIPConenctionState(Enum):
    NOT_CONNECTED = 0
    CONNECTING = 1
    CONNECTED = 2


class LIPSocket:
    """A socket that reads and writes lip protocol."""

    def __init__(self, reader, writer):
        self._writer = writer
        self._reader = reader

    async def async_readline(self, timeout=SOCKET_TIMEOUT):
        buffer = await asyncio.wait_for(self._reader.readline(), timeout=timeout)
        if buffer == b"":
            return None

        return buffer.decode("UTF-8")

    async def async_readuntil(self, seperator, timeout=SOCKET_TIMEOUT):
        buffer = await asyncio.wait_for(
            self._reader.readuntil(seperator.encode("UTF-8")), timeout=timeout
        )
        if buffer == b"":
            return None

        return buffer.decode("UTF-8")

    async def async_write_command(self, text):
        self._writer.write(text.encode("UTF-8") + b"\r\n")
        await self._writer.drain()

    def close(self):
        """Cleanup when disconnected."""
        self._writer.close()

    def __del__(self):
        """Cleanup when the object is deleted."""
        self._writer.close()


aiolip/aiolip.py
import asyncio
import re
import socket
from dataclasses import dataclass
from enum import Enum

LIP_USERNAME = "lutron"
LIP_PASSWORD = "integration"
LIP_PORT = 23

LIP_RESPONSE_RE = re.compile("~([A-Z]+),([0-9.]+),([0-9.]+),([0-9.]+)")

LIP_PROTOCOL_LOGIN = "login: "
LIP_PROTOCOL_PASSWORD = "password: "
LIP_PROTOCOL_GNET = "GNET> "

LIP_QUERY_CHAR = "?"
LIP_ACTION_CHAR = "#"

LIP_BUTTON_PRESS = 3
LIP_BUTTON_RELEASE = 4

LIP_KEEP_ALIVE = "?SYSTEM,10"

CONNECT_TIMEOUT = 10
SOCKET_TIMEOUT = 10

LIP_READ_TIMEOUT = 60
LIP_KEEP_ALIVE_INTERVAL = 60


@dataclass
class LIPMessage:
    """Class for LIP messages."""

    mode: str
    integration_id: int
    action_number: int
    value: float


class LIPConnectionStateError(Exception):
    """An exception to represent a conneciton state error."""

    def __str__(self) -> str:
        """Return string representation."""
        return "Lutron Integration Protcol is not connected."


class LIPProtocolError(Exception):
    """An exception to represent a protocol error."""

    def __init__(self, received, expected):
        self.received = received
        self.expected = expected

    def __str__(self) -> str:
        """Return string representation."""
        return f"Lutron Protocol Error received=[{self.received}] expected=[{self.expected}]"


class LIPConenctionState(Enum):
    NOT_CONNECTED = 0
    CONNECTING = 1
    CONNECTED = 2


class LIPSocket:
    """A socket that reads and writes lip protocol."""

    def __init__(self, reader, writer):
        self._writer = writer
        self._reader = reader

    async def async_readline(self, timeout=SOCKET_TIMEOUT):
        buffer = await asyncio.wait_for(self._reader.readline(), timeout=timeout)
        if buffer == b"":
            return None

        return buffer.decode("UTF-8")

    async def async_readuntil(self, seperator, timeout=SOCKET_TIMEOUT):
        buffer = await asyncio.wait_for(self._reader.readuntil(seperator.encode("ASCII")), timeout=timeout)
        if buffer == b"":
            return None

        return buffer.decode("UTF-8")

    async def async_write_command(self, text):
        self._writer.write(text.encode("ASCII") + b"\r\n")
        await self._writer.drain()

    def __del__(self):
        """Cleanup when the object is deleted."""
        self._writer.close()


class LIP:
    """Async class to speak LIP."""

    def __init__(self):
        """Create the LIP class."""
        self._connection_state = LIPConenctionState.NOT_CONNECTED
        self._lip = None
        self._host = None
        self._disconnect_event = None
        self._subscriptions = []
        self.loop = None

    def connection_state(self):
        """The current state of the connection."""
        return self._connection_state

    async def async_connect(self, server_addr):
        """Connect to the bridge via LIP."""
        self.loop = asyncio.get_event_loop()
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(
                server_addr,
                LIP_PORT,
                family=socket.AF_INET,
            ),
            timeout=CONNECT_TIMEOUT,
        )
        self._lip = LIPSocket(reader, writer)
        _verify_expected_response(await self._lip.async_readuntil(" "), LIP_PROTOCOL_LOGIN)
        await self._lip.async_write_command(LIP_USERNAME)
        _verify_expected_response(await self._lip.async_readuntil(" "), LIP_PROTOCOL_PASSWORD)
        await self._lip.async_write_command(LIP_PASSWORD)
        _verify_expected_response(await self._lip.async_readuntil(" "), LIP_PROTOCOL_GNET)

        self._connection_state = LIPConenctionState.CONNECTED
        self._disconnect_event = asyncio.Event()
        self._host = server_addr

    async def _async_reconnect(self):
        """Reconnect after disconnected."""
        await self.async_connect(self._host)

    async def async_stop(self):
        """Disconnect from the bridge."""
        self._disconnect_event.set()
        if self._keep_alive_task:
            self._keep_alive_task.cancel()
            self._keep_alive_task = None

    async def _keep_alive_or_reconnect(self):
        """Start keep alive."""
        try:
            await self._async_send_keep_alive()
        except ConnectionResetError:
            await self._async_reconnect()

        if self._disconnect_event.is_set():
            return

        self._keep_alive_task = self.loop.call_later(
            LIP_KEEP_ALIVE_INTERVAL, self._keep_alive_or_reconnect
        )

    async def async_start(self):
        """Start interacting with the bridge."""

        self._verify_connected()
        await self._keep_alive_or_reconnect()

        while True:
            disconnect_task = asyncio.create_task(self._disconnect_event.wait())
            read_task = asyncio.create_task(self._lip.async_readline())
            try:
                await asyncio.wait(
                    (read_task, disconnect_task), timeout=LIP_READ_TIMEOUT
                )
            except asyncio.TimeoutError:
                continue

            if self._disconnect_event.is_set():
                return

            match = read_task.result().match(LIP_RESPONSE_RE)
            if match:
                try:
                    message = LIPMessage(
                        match.group(1).decode("utf-8"),
                        int(match.group(2)),
                        int(match.group(3)),
                        float(match.group(4)),
                    )
                    for callback in self._subscriptions:
                        callback(message)
                except ValueError:
                    pass

    async def query(self, mode, integration_id, action, *args):
        """Query the bridge."""
        await self._async_send_command(LIP_QUERY_CHAR, [mode, integration_id, action, *args])

    async def action(self, mode, integration_id, action, *args):
        """Do an action on the bridge."""
        await self._async_send_command(LIP_ACTION_CHAR, [mode, integration_id, action, *args])

    async def _async_send_command(self, protocol_header, cmd):
        self._verify_connected()
        request = ",".join(cmd)
        await self._lip.async_write_command(f"{protocol_header}{request}")

    def subscribe(self, callback):
        """Subscribe to lip events."""

        def _unsub_callback():
            self._subscriptions.remove(callback)

        self._subscriptions.append(callback)
        return _unsub_callback

    def _verify_connected(self):
        """Throw if we are disconnected."""
        if self._connection_state != LIPConenctionState.CONNECTED:
            raise LIPConnectionStateError


def _verify_expected_response(expected, received):
    if not expected.startswith(received):
        raise LIPProtocolError(expected, received)


async def main():
    lip = LIP()

    await lip.async_connect("192.168.209.70")

    def message(msg):
        import pprint

        pprint.pprint(msg)

    lip.subscribe(message)
    await lip.async_start()


if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
